core/shared/src/main/scala/zio/ZIO.scala

```diff
@@ -2098,11 +2098,46 @@ private[zio] trait ZIOFunctions extends Serializable {
    *
    * For a sequential version of this method, see `foreach`.
    */
-  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] =
-    as.foldRight[ZIO[R, E, List[B]]](effectTotal(Nil)) { (a, io) =>
-        fn(a).zipWithPar(io)((b, bs) => b :: bs)
-      }
-      .refailWithTrace
+  final def foreachPar[R, E, A, B](as: Iterable[A])(fn: A => ZIO[R, E, B]): ZIO[R, E, List[B]] = {
+    val n         = as.size
+    val resultArr = new Array[Any](as.size)
+    val resultList: ZIO[R, E, List[B]] = for {
+      parentId <- ZIO.fiberId
+      latch    <- CountdownLatch.make(n)
+      failed   <- Promise.make[Nothing, Int]
+      cause    <- Ref.make[Cause[E]](Cause.empty)
+      fibers <- ZIO.foreach(as.zipWithIndex) {
```

Comment:
`zipWithIndex` will create an intermediate collection and together with `ZIO.foreach` there would be two traversals only on this line. I'm wondering if we can use a `view` here (I'm aware that it's not a very popular choice) or the stdlib `Stream/LazyList` (but we will have deprecation issues here).

As an alternative, can we maybe think of a way to force the original `Iterable` to a concrete data type (maybe even a `Vector` for its constant time `size` method) and manually construct the indices that would otherwise require a whole traversal with `zipWithIndex`.

Reply:
I think any way of getting rid of the extra traversal would be fine here as it is fairly hidden away. I'll experiment.

If altering the signature is on the table, we could make both the parameter and return types IndexedSeq which would allow constant time size and avoid having to repack into a list at the end. It does make it quite a bit more inflexible though. What do you think?

Reply:
It is definitely fine to use impure code in the implementation, as long as it doesn't leak outside or is observable in any way.

I don't know about altering the method signature. Let's try using an IndexedSeq (Vector) inside the method (.toVector) and see where we can go from there.

Reply:
Oddly using a view made it go slower!

The new commit I just pushed reduces the traversals and improves the performance by a small amount. I also tried a completely vector/array based approach as you suggested (which I've put in another branch for comparison). Slightly surprisingly, it comes out slightly slower in the benchmark (though the difference is not large).

Reply:
Views aren't exactly know for their amazing performance, so I cannot say that I'm surprised.